Bluetooth: ISO: Add seq_num and ts to bt_iso_chan_send

Thalley opened the pull request:
Add two new parameters to bt_iso_chan_send:
sn: The packet sequence number which shall be incremeted per SDU interval.
ts: An optional timestamp value used to synchronize SDUs.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

fixes https://github.com/zephyrproject-rtos/zephyr/issues/43859

Thalley committed:
Bluetooth: ISO: Add sn and ts to bt_iso_chan_send

Add two new parameters to bt_iso_chan_send:
sn: The packet sequence number which shall be incremeted
per SDU interval.
ts: An optional timestamp value used to synchronize SDUs.

The sequence number in the API uses a 32-bit value even though
the sequence number for the HCI command is 16-bit. This is to
properly handle wrapping of sequence numbers, which is much
easier to do with additional bits allocated.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: ISO: Rename ISO_INTERVAL_MIN/MAX

Rename the BT_ISO_INTERVAL_MIN/MAX to
BT_ISO_SDU_INTERVAL_MIN/MAX to avoid confusing this
the the ISO interval which is different from the SDU
interval.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented:
Added DNM as I am awaiting response on https://www.bluetooth.org/errata/errata_view.cfm?errata_id=18552 to see if this is truly the right way of doing it, or if the spec is unclear or unintentionally creating a bad requirement from the host.

Thalley commented:
Rebased to solve merge conflicts

Thalley commented:
Rebased to fix CI

Thalley commented:
Rebased to solve merge conflicts

Thalley commented:
Rebased to solve merge conflicts

Casper-Bonde-Bose commented on include/zephyr/bluetooth/iso.h:
```diff
@@ -86,6 +86,18 @@ extern "C" {
+ * When this value is used as the sequence number in bt_iso_chan_send() the
+ * sequence number range starts again from 0.
+ */
+#define BT_ISO_SEQUENCE_NUMBER_RESET 0xFFFFU
```
As the sequence number is a 16 bit unsigned value 0xffff is a valid sequence number to use (will occur every 11 minutes for 10ms frames).
Maybe use a 32 bit value for reset?
UPDATE: I see that the reset value is actually not used to force-reset to 0, but is accepted as a valid value - clever! - no change required:-)

Thalley replied:
Hmm, one issue with this approach is the following scenario:

The host sends an SDU with SN `0xFFFE`, and the next SDU is 3 intervals later, and would thus be sent with SN `0x0001`. 

The approach with resetting the sequence number using `0xFFFF` won't work in that scenario. 

So I think that we either need to come up with some way of determining which SNs we would consider value (e.g. larger than current or perhaps lower than e.g. 1000), or we need to remove the verification of the sequence number all together and just send to the controller regardless.

Casper-Bonde-Bose replied:
I think it would work.
If the app do not reset `0x0001` is larger than `0xFFFE` (as it wraps) - hence that is valid (if the check handles wraps of cause - I did not consider that during the review :-) )
If the app makes a call to reset the SN (`0xFFFF`) and then after 3 SDU intervals sends with the SN `0x0001` it should be accepted.

But yes for a uint16_t you would need to define what a larger value is. A difference of 1000 is a good number, but perhaps use half the range of uint16_t? But then again that basically disables the check and you might as well remove it. In the case you pause transmission for 6+ minutes the API call would fail even though you use a correct SN.

One alternative would be to have the API and check(iso_conn->iso.seq_num) use a 32 bit values - then you would not have the issue to determine whether the SN wrapped or not.

Thalley replied:
> If the app makes a call to reset the SN (0xFFFF) and then after 3 SDU intervals sends with the SN 0x0001 it should be accepted.

The current implementation still uses 0xFFFF as a valid value, and would attempt to send a SDU, which would be unwanted in that case. To reset the counter without sending anything, we should need to use at least 17 bits and and have a "reset only" value. 

> But yes for a uint16_t you would need to define what a larger value is. A difference of 1000 is a good number, but perhaps use half the range of uint16_t? But then again that basically disables the check and you might as well remove it. In the case you pause transmission for 6+ minutes the API call would fail even though you use a correct SN.

When playing around with the updated shell approach, i.e. calculating the next SN instead of doing it by timer, it occurred to me how fast the SNs would wrap (approximately every 16 seconds at 255us). There is no way of handling sporadic sequence numbers from the shell as you can easily send SN 20000 and then the next be at SN 5000 when using something like the shell. Granted that the shell probably shouldn't be the thing that decides how we handle this, but on the other hand, given the issue with the SN for the peripheral and updating the SN every 255us, we still need something to work for the shell. 

> One alternative would be to have the API and check(iso_conn->iso.seq_num) use a 32 bit values - then you would not have the issue to determine whether the SN wrapped or not.

Until it's time to wrap the 32-bit value ;) But no, this could be a decent work around, as the wrapping timing would go from 65 seconds to 1193 hours (at 10ms interval), and we could then allow for values than the current once the current value reaches `0x10000` (i.e. larger than a uint16_t.

Casper-Bonde-Bose commented on subsys/bluetooth/shell/iso.c:
```diff
@@ -33,7 +49,41 @@ static void iso_recv(struct bt_iso_chan *chan, const struct bt_iso_recv_info *in
+			cis_timer_initialized = true;
+		}
+
+		k_work_schedule(&cis_sn_work, K_USEC(cis_sdu_interval_us));
```
The minimum SDU interval is 255us but the tick timer is typically between 1ms and 10ms - hence you will not have a stable timer firing every 255us - at least not when the system is busy. I think the only stable way to increment this counter is to take a timestamp here and calculate the increment every time send() is called.

Thalley replied:
That's a good idea, basically calculating the (minimum) SN increment instead of continuously incrementing it by 1. Will update.

Thalley replied:
Hmm, implemented it and it seems to work correctly. However, when we are still using 255us as the increment factor, then the values are incrementing quite fast (in fact the SN wraps every ~16 second), which makes it damn near impossible to use this with the SN verification that I implemented... 

Another solution would be to, for the ISO peripheral, to have a command that tells the shell what the SDU interval is and use that, and avoid the 255us increment altogether.

Casper-Bonde-Bose reviewed (dismissed):
Comment in shell sample application is optional to fix IMO as it is a sample application.

Thalley commented:
Removing DNM as we have a workaround for the erratum. 

Updated the shell to calculate the next sequence number instead of incrementing it. 

Updated the sequence number to use uint32_t instead of uint16_t.

Casper-Bonde-Bose commented on subsys/bluetooth/host/iso.c:
```diff
@@ -765,13 +765,39 @@ int bt_iso_chan_send(struct bt_iso_chan *chan, struct net_buf *buf)
+	 * the sequence number which should provide amble room to handle
+	 * applications that does not send an SDU every interval.
+	 */
+	if (sn <= iso_conn->iso.seq_num &&
+	    iso_conn->iso.seq_num < BT_ISO_MAX_SEQ_NUM) {
+		BT_DBG("Invalid sn %u - Shall be > than %u",
+		       sn, iso_conn->iso.seq_num);
+		return -EINVAL;
+	}
+
+	iso_conn->iso.seq_num = sn;
```
Nice, so even when the counter wraps after a minimum of 248 days, it will still work. Great for sensor data streaming, that is intended to run for years.

Thalley replied:
Yes. Of course this also means that an application can do

```c
sn = 0x00001;
iso_send(sn); /* controller gets 0x0001 */
sn = 0x10001;
iso_send(sn); /* controller gets 0x0001 */
sn = 0x20001;
iso_send(sn); /* controller gets 0x0001 */
```
So it is still possible to abuse, but without any core spec specified value wrapping protocol this is unavoidable.

Casper-Bonde-Bose reviewed (dismissed):
LGTM
Minor suggestion to add info about the unit of the timestamp.

Casper-Bonde-Bose reviewed (dismissed):
LGTM

Thalley commented:
Rebased to solve merge conflicts

Casper-Bonde-Bose reviewed (dismissed):
LGTM

asbjornsabo commented on samples/bluetooth/central_iso/src/main.c:
```diff
@@ -60,13 +62,13 @@ static void iso_timer_timeout(struct k_work *work)
 	}
 
-	k_work_schedule(&iso_send_work, K_MSEC(1000));
+	k_work_schedule(&iso_send_work, K_USEC(interval_us));
```
Timing is changed with a factor of ten here - is that intentional?

Thalley replied:
Timing was changed to match the SDU interval, which is intentional as we want to use this callback to increment the SN per SDU interval

asbjornsabo commented on subsys/bluetooth/host/iso.c:
```diff
@@ -716,13 +716,39 @@ int bt_iso_chan_send(struct bt_iso_chan *chan, struct net_buf *buf)
+	/* Once the stored seq_num reaches the maximum value sendable to the
+	 * controller (BT_ISO_MAX_SEQ_NUM) we will allow the application to wrap
+	 * it. This ensures that up to 2^32-1 SDUs can be sent without wrapping
+	 * the sequence number which should provide amble room to handle
```
"amble"-> "ample"

asbjornsabo commented on subsys/bluetooth/host/conn_internal.h:
```diff
@@ -126,8 +126,13 @@ struct bt_conn_iso {
-	uint16_t seq_num;
+	/** @brief 16-bit sequence number that shall be incremented per SDU interval
+	 *
+	 *  We store it as 32-bit to handle wrapping: Only once the value has
```
Nitpicking: Remove "we"?
"We store it" -> "Stored"
"will we allov [...]" -> "will values less than the current be allowed again"

asbjornsabo commented on subsys/bluetooth/host/conn_internal.h:
```diff
@@ -126,8 +126,13 @@ struct bt_conn_iso {
 #if defined(CONFIG_BT_ISO_UNICAST) || defined(CONFIG_BT_ISO_BROADCASTER)
-	/** 16-bit sequence number that shall be incremented per SDU interval */
-	uint16_t seq_num;
+	/** @brief 16-bit sequence number that shall be incremented per SDU interval
```
Is it a good thing to have this "shall" here?
This is the struct, is this the right place to mandate behavior?  And if we do, should it say who is responsible to increment it?
Propose to end sentence after sequence number".

Thalley replied:
I think it's fine to keep it in this (internal) struct, as that will serve as a comment for developers if they attempt to change how the `seq_num` is handled

asbjornsabo commented on subsys/bluetooth/shell/iso.c:
```diff
@@ -233,6 +290,16 @@ static int iso_accept(const struct bt_iso_accept_info *info,
 
 	*chan = &iso_chan;
 
+	/* As the peripheral host we do no know the SDU interval, and thus we
```
"no" -> "not"

asbjornsabo commented on subsys/bluetooth/shell/iso.c:
```diff
@@ -233,6 +290,16 @@ static int iso_accept(const struct bt_iso_accept_info *info,
 	*chan = &iso_chan;
 
+	/* As the peripheral host we do no know the SDU interval, and thus we
+	 * cannot find the proper interval og incrementing the packet
```
"og" -> "of"?  
or  -> "for"?

asbjornsabo reviewed (changes requested):
A couple of minor spelling corrections, and some questions for my understanding.

Thalley commented on samples/bluetooth/central_iso/src/main.c:
```diff
@@ -23,6 +23,8 @@ static void start_scan(void);
 static struct k_work_delayable iso_send_work;
 static struct bt_iso_chan iso_chan;
+static uint32_t sn;
+static uint32_t interval_us = 100U * USEC_PER_MSEC; /* 100 ms */
```
It is the SDU interval :)

Thalley commented on samples/bluetooth/central_iso/src/main.c:
```diff
@@ -238,7 +242,7 @@ void main(void)
 	param.framing = 0;
 	param.latency = 10; /* ms */
-	param.interval = 10 * USEC_PER_MSEC; /* us */
+	param.interval = interval_us; /* us */
```
Hmm, I actually don't remember, nor do I see any reason, to change from 10ms to 100ms. Will change back to 10ms.

Thalley commented on subsys/bluetooth/host/iso.c:
```diff
@@ -716,13 +716,39 @@ int bt_iso_chan_send(struct bt_iso_chan *chan, struct net_buf *buf)
+
+		hdr = net_buf_push(buf, sizeof(*hdr));
+		hdr->ts = ts;
+		hdr->data.sn = sys_cpu_to_le16((uint16_t)sn);
```
Because one is a `bt_hci_iso_data_hdr` and the other is a `bt_hci_iso_ts_data_hdr` :) 

They are defined as 

```c
struct bt_hci_iso_data_hdr {
	uint16_t sn;
	uint16_t slen;
} __packed;

struct bt_hci_iso_ts_data_hdr {
	uint32_t ts;
	struct bt_hci_iso_data_hdr data;
} __packed;
```

Casper-Bonde-Bose reviewed (dismissed):
LGTM

Thalley committed:
Bluetooth: ISO: Rename `sn` to `seq_num`

Rename the `sn` field(s) to `seq_num` to be more readable
and clear what the value covers.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented:
Added a new commit to rename `sn` to `seq_num` as per request from @cvinayak: https://github.com/zephyrproject-rtos/zephyr/pull/45436#discussion_r887561157

Thalley commented:
Rebased to solve merge conflicts

sjanc commented on include/zephyr/bluetooth/iso.h:
```diff
@@ -222,7 +228,7 @@ struct bt_iso_recv_info {
 
 	/** ISO packet sequence number of the first fragment in the SDU */
-	uint16_t sn;
+	uint16_t seq_num;
```
isn't this an API change?

Thalley replied:
It is, but the ISO API is still considered experimental, so we can do changes like this without worrying too much :)

sjanc replied:
ok, I've only checked CONFIG_BT_ISO_UNICAST and BT_ISO_BROADCAST and those aren't market as experimental :)
